lexicon/providers/misaka.py

```diff
@@ -0,0 +1,229 @@
+        self.domain_id = self.domain
+
+    # Create or update a record.
+    def _create_record(self, rtype, name, content):
+        name = self._relative_name(name)
+        print(name)
+
+        identifier = f'{self.domain_id}/recordsets/{name}/{rtype}'
+        endpoint = f'/zones/{identifier}'
+        ttl = self._get_lexicon_option('ttl')
+        record = {'value': content}
+
+        existing_recordset = self._get_recordset(name, rtype)
+        if existing_recordset:
+            # append if returned recordsets doesn't include the record
+            if not _recordset_has_record(existing_recordset, content):
+                existing_recordset['records'].append(record)
+                existing_recordset['ttl'] = ttl
+                self._put(endpoint, existing_recordset)
+                LOGGER.debug('recordset exists, appending: %s', identifier)
+            # update ttl if returned ttl doesn't match
+            elif existing_recordset['ttl'] != ttl:
+                existing_recordset['ttl'] = ttl
+                self._put(endpoint, existing_recordset)
+                LOGGER.debug('recordset exists, updating ttl: %s', identifier)
+        else:
+            recordset = {
+                'ttl': ttl,
+                'filters': [],
+                'records': [record],
+            }
+            self._post(endpoint, recordset)
+            LOGGER.debug('create_record: %s', identifier)
+
+        return True
+
+    def _list_records(self, rtype=None, name=None, content=None):
+        params = {}
+        if name:
+            name = self._relative_name(name)
+            params['name'] = name
+
+        payload = self._get(f'/zones/{self.domain_id}/recordsets', query_params=params)
+        records = []
+        for recordset in payload['results']:
+            if not recordset['records']:
+                continue
+
+            if name and recordset['name'] != name:
+                continue
+
+            if rtype and recordset['type'] != rtype:
+                continue
+
+            processed_recordset = {
+                'name': self._full_name(recordset['name']),
+                'type': recordset['type'],
+                'ttl': recordset['ttl'],
+
+                # id is useless here as we don't have record ID exposed to customers
+                'id': '{}/recordsets/{name}/{type}'.format(self.domain_id, **recordset),
+            }
+
+            for record in recordset['records']:
+                processed_record = {'content': record['value']}
+                processed_record.update(processed_recordset)
+                records.append(processed_record)
+
+        LOGGER.debug('list_records: %s', records)
+        return records
+
+    # Create or update a record.
+    def _update_record(self, identifier, rtype=None, name=None, content=None):
+        # use relative_name only
+        if name:
+            name = self._relative_name(name)
+
+        new_identifier = f'{self.domain_id}/recordsets/{name}/{rtype}'
```

Comment:
So misaka.io does not implement identifiers in their API, and you need to generate these IDs on Lexicon side.

Storing the parameters in the identifier to retrieve the record associated is a smart move. However, it has some drawbacks on CLI usability, since from user side you need to deal with a complex string instead of usual simple identifiers expected here (serie of number or some letters).

I personally prefer to use another approach in this situation. I calculate dynamically the identifier to use from an hash-baset approach. Typical code is:

https://github.com/AnalogJ/lexicon/blob/cc985166aead697727e56517c58061693ea087ce/lexicon/providers/godaddy.py#L227-L233

Then, when an identifier is provided to update or delete a record, I get all records of the zone, calculate the identifier for each of them, and select the one corresponding to the identifier provider.

This way, you have an identifier that is not exposing some internals of the fetching logic, and keep the identifier simple (here 8 characters).

It can be a problem however if fetching all records is quite slow for the DNS API.

What do you think of this approach?

Reply:
As a comprehensive approach, we can remove `{self.domain_id}/recordsets/` part from the identifier.
Since `domain` is a mandatory argument we don't need it in identifier.

Local-calculated dynamic identifier does make it easier when typing manually on CLI but its cost is a little bit higher and I would like to avoid it if it's not required.

Reply:
So it would leave only `{name}/{type}`. Ok I think we can work from that. I propose you the following:
* we always use the `relative_name`, not the `full_name` for the `name` parameter, since `full_name` can be processed from `relative_name` using the domain if needed.
* we encode the `{name}/{type}` in Base64URL (`base64.urlsafe_b64encode()`) to get the actual identifier for maximum portability
* as a bonus we can strip the trailing padding `=` characters with something like https://gist.github.com/cameronmaske/f520903ade824e4c30ab, but I think it not that necessary